unison-cli/src/Unison/Util/Servant/Client.hs

```diff
@@ -0,0 +1,99 @@
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Unison.Util.Servant.Client where
+
+import Data.Generics.Sum
+import Data.Proxy
+import qualified Data.SOP as SOP
+import Servant.API
+import qualified Servant.API.UVerb as UVerb
+
+-- | This typeclass allows us to handle Servant Union responses in a type-safe way.
+--
+-- Servant 'UVerb' responses look like this:
+--
+-- @@
+-- UVerb 'POST '[JSON] '[WithStatus 204 (), WithStatus 404 (NeedDependencies Hash), WithStatus 412 HashMismatch]
+-- @@
+--
+-- Each response is tagged with its status code, and each response type can be different.
+--
+-- We can't pack all the responses into a single type because servant needs the status code at the
+-- type-level.
+--
+-- servant-client treats the return type of a method like this as:
+--
+-- @@
+-- (Union '[WithStatus 204 (), WithStatus 404 (NeedDependencies Hash), WithStatus 412 HashMismatch])
+-- @@
+--
+-- Which is an SOP Union under the hood which we need to unpack.
+--
+-- We can pattern match on it, but it ends up very complex, e.g.
+--
+-- @@
+-- case result of
+--   (SOP.Z (SOP.I (WithStatus a))) -> UpdatePathSuccess
+--   SOP.S (SOP.Z (SOP.I (WithStatus a))) -> UpdatePathHashMismatch a
+--   SOP.S (SOP.S (SOP.Z (SOP.I (WithStatus a)))) -> UpdatePathMissingDependencies a
+-- @@
+--
+-- And the error messages are pretty atrocious when we mess it up.
+--
+-- Instead, we can leverage the 'AsType' class from `generic-lens`, which allows us to
+-- automatically inject subtypes into the appropriate sum-type constructor for us.
+--
+-- E.g. if we have:
+--
+-- @@
+-- data UpdatePathResponse
+--   = UpdatePathSuccess
+--   | UpdatePathHashMismatch HashMismatch
+--   | UpdatePathMissingDependencies (NeedDependencies Hash)
+-- @@
+--
+-- Then @injectTyped (hashMismatch :: HashMismatch)@ gives us: @UpdatePathHashMismatch
+-- hashMismatch@
+-- And @injectTyped ()@ gives us @UpdatePathSuccess@
+--
+-- We can then use this as a constraint with 'foldMapUnion' to collapse the set of
+-- responses down to exactly the UpdatePathResponse type!
+--
+-- generic-lens will now provide us with really nice error messages if we're missing a
+-- constructor for a given response, e.g.:
+--
+--  • The type UpdatePathResponse does not contain a constructor whose field is of type HashMismatch
+--  • In the expression: liftHandlerUnion updatePathHandler
+class HasConstructor typ member where
```

Comment:
See the included docs, overall I'm a bit sad that I needed to do something like this, but am happy with the result.

These are the only docs I could find on handling servant responses [docs](https://docs.servant.dev/en/stable/cookbook/uverb/UVerb.html), which are written by some previous co-workers of mine at Wire.

The approach they use is a little simpler, but doesn't seem to actually maintain a type-level guarantee that we properly handle every response type (it's all `Maybe` typed, so you wouldn't get any compile-time errors).

I'm happy with where it ended up though, largely this module can be left completely alone and we can just use `liftHandlerUnion` in the HTTP module and it all "Just works™" and gives surprisingly good error messages.